packages/nimble-components/src/rich-text/specs/mention-hld.md

```diff
@@ -0,0 +1,721 @@
+
+## Links To Relevant Work Items and Reference Material
+
+-   [Rich Text Editor and Viewer README](./README.md)
+-   [Comments `@mention` mockup](https://www.figma.com/file/Q5SU1OwrnD08keon3zObRX/SystemLink-orig?type=design&node-id=7248-114254&mode=design&t=y3JtM3aT77Aw0xjK-0)
+-   [Mention users in comments - Requirement doc](https://dev.azure.com/ni/DevCentral/_git/Skyline?path=/docs/design-documents/Platform/Requirements/Mention-users-in-comments.md&_a=preview)
+-   [Work Item for @mention in comments in AzDo](https://dev.azure.com/ni/DevCentral/_workitems/edit/2464517)
+
+## Design
+
+### `nimble-rich-text-editor`
+
+The rich text editor includes this feature for tagging or mentioning a user by inserting the **"@"** symbol within the editor.
+This action triggers a dropdown list, allowing users to select a user. To provide the rich text editor component with the
+necessary user information for populating this dropdown, clients can pass the user details through the configuration elements outlined below:
+
+```HTML
+<nimble-rich-text-editor>
+    <nimble-rich-text-mention-users pattern="http://users/.*">
+        <nimble-mapping-mention-user
+            mention-url="http://users/user-id-1"
+            display-name="John Doe"
+        ></nimble-mapping-mention-user>
+        <nimble-mapping-mention-user
+            mention-url="http://users/user-id-2"
+            display-name="Alice Smith"
+        ></nimble-mapping-mention-user>
+        <nimble-mapping-mention-user
+            mention-url="http://users/user-id-3"
+            display-name="Bob Jones"
+        ></nimble-mapping-mention-user>
+    </nimble-rich-text-mention-users>
+</nimble-rich-text-editor>
+```
+
+The configuration element, `nimble-rich-text-mention-users`, consists of mapping elements that specify both the content to display
+in the dropdown list (i.e., the `display-name`) and the data to store in the markdown when extracting content from the editor (i.e., the `mention-url`).
+These details are subsequently transformed into a map or an object, which is used to populate the options within the shadow root
+for the dropdown list of items. This component uses the [`pattern`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern)
+attribute to acquire the regular expression pattern that corresponds to the URLs of users within the specific domain of the client application.
+
+The `nimble-mapping-mention-user` element contains individual user information within its elements. Typically, the `mention-url` attribute
+includes either the user ID or a value stored in the backend database. It's important that the `mention-url` adheres to a valid URL format,
+often used to link the user ID with the hosting domain. In cases where there's no specific URL for each user within the domain,
+they can alternatively insert the user ID in a valid URL format such as `user:user-id` and provide the appropriate pattern for it.
```

Comment:
The term valid URL I meant it is a valid [Autolink format in CommonMark](https://spec.commonmark.org/0.30/#autolink). I updated the sentence to explicitly mention it should be in Autolink format rather than a valid URL that might create confusion.

Since we are planning to give higher priority to the `mention` node than to the `autolink` mark, even non-HTTP and non-HTTPS links will first be processed by the `mention` node. Only if the given href doesn't match the specified pattern will it be considered for the `autolink` mark. The restriction for HTTP/HTTPS in `autolink` has not been removed and will render as plain text if the autolink doesn't match with the `pattern` or it is not HTTP/HTTPS.

Reply:
Only concern I have is that the child configuration elements are able to come in dynamically. So after a set Markdown call happens, a couple seconds later a mention config could be added, so what was previously an unmatched link can become a matched link and vice versa. That might just be testing requirements.

Reply:
This could be one of the corner cases to consider. However, if the mention config is not added or added without a pattern then the `@mention` feature is disabled in the editor. The actual recommendation for the client is to load only the mapping elements dynamically and not the config element. So in that case, if they add the config element dynamically it might create a conflict with the API we have for setting the markdown through a method.

I think to handle the case we must maintain a markdown string state inside an editor for every `setMardown(mardown)` is called by the client and after a couple of seconds if the child config elements changed, we can call the `setMardown(mardown)` again with the private markdown string we maintain in the editor and re-populate by matching the href. In the viewer, we can update the view based on the child config element change and get the markdown string from the public `markdown` property we have.

Reply:
With option 1, we see some difficulties in extracting the current state from the editor to perform the getMarkdown + setMarkdown scenario, here are the problems,

1. If we preserve links only in the markdown and render them as plain text if not a supported scheme, then the rendered node in the editor will be a paragraph tag. And, during the serialization, those paragraph nodes will converted as plain text rather than an autolink.

    For example, if the client sets a markdown `<user:123>` but without a config child element. In the editor, it will be rendered as plain text within the paragraph node. After a few seconds, the config element is added with the pattern as `user:.*` then when performing `getMardown()`, the serializer will detect it as plain text and not a autolink because the current state in ProseMirror will have only paragraph node and not a link node.

    So the only solution we could come up with handling this case to work as you suggested, the unsupported scheme to also be rendered as a link node but change the appearance of the link to look and interact like plain text by styling in CSS. So that when we `getMarkdown()`, the ProseMirror identifies the unsupported scheme also as a link, and when `setMarkdown()` it matches with the pattern and renders as mention node.

2. If we should lift the restrictions from the rich text components and render all scheme links as a link in the editor, then the problem is solved for this case without even writing any CSS workarounds mentioned above. But all the scheme links by default will render as links and people can even render the same link in the viewer too (which is clickable) because we will be removing the restriction from the parser that is shared between editor and viewer.

3. So this leads us to option 2 of your suggestion by restricting https/http links in `mention-href` of the mapping elements.

Please let us know whether option 1 can be implemented to preserve the link in the markdown and appear as plain text in the editor and viewer for unsupported scheme.

Reply:
We cannot use just CSS to disable the link. There will be weird interactions with `pointer-events: none` (around selection behavior, etc) and CSS does not impact keyboard and accessibility either. We would need additional hacks like changing the `role` attribute and the `tabindex`. I don't even know if all those hacks are sufficient either.

The correct way to disable a link is to remove the `href` attribute from the `<a>` tag. By removing the `href` attribute the link will render as plain text correctly on it's own. The best option would be if we can avoid emitting the `href` attribute on the anchor tag or emit a different attribute instead of href we can detect like `data-nimble-rich-text-disabled-href`.

Reply:
Yes! It makes sense to not have `href` for the unsupported scheme, render it as plain text, and still preserve all the schemes in autolink format in markdown. So even when changing the config element dynamically we can update it to a mention node when the pattern matches. I have updated the spec on allowing all the schemes in the editor and viewer for links.